### ISSUE TITLE
Incorrect variable period timestep when timestep gets large

### ISSUE THREAD

**Issue by Kappie** [glotzerlab/hoomd-blue#643]
I am experiencing **incorrect** behavior when I use the following function to generate a variable period timestep.

```python
def log_period(block_size, base=2):
    def period(n):
        max_exp = int(math.log(block_size, base)) + 1
        n_mod = n % max_exp
        n_div = n // max_exp
        if n_mod == 0:
            return n_div * block_size 
        else:
            return n_div * block_size + base**(n_mod - 1)

    return period
```

I am using `log_period(2 ** 18, base=2)` from the Python side to save a logarithmically spaced trajectory. To illustrate, `log_period(2**3, base=2)` would return a function that generates the sequence `0, 1, 2, 4, 8, 9, 10, 12, 16, 17, 18, ...`.

Between timestep 16726904 and 16779322 I get the error `*Warning*: Variable period returned a value equal to the current timestep. Increasing by 1 to prevent inconsistencies` for the first time, and many times after that. 

In the actual saved trajectory, the first time step that is wrong is 671088**67** (it should be one more, 671088**68**). This table shows what is happening:

| m_n | actual step | period(m_n) in Python (expected) |
| --- | ----------- | -------------------------- |
| 4864 | 671088**64** | 671088**64** | 
| 4865 | 671088**65** | 671088**65** |
| 4866 | 671088**66** | 671088**66** |
| 4867 | 671088**67** (??!) | 671088**68**  |
| 4868 | 671088**72** | 671088**72** |

For what it's worth, the error happens after exactly 256 "blocks", where a block is [0, 1, 2, ..., 2**17] mod 2 ** 18.

It seems to me that quite possibly the relevant code is in System.h line 216.

```C++
pybind11::object pynext = m_update_func(m_n);
int next = (int)pybind11::cast<float>(pynext) + m_created_tstep;
``` 

Here, the output from the Python period function first gets casted to float, and then to int.

Thanks for a great library, and if there is any additional information I can supply, I will be happy to provide it.

**Comment by joaander** [comment]
Thanks for the detailed bug report. This issue will be resolved in HOOMD v3.0 with the newly implemented `Trigger` API. A `Trigger` takes a 64-bit unsigned int timestep as input and returns true or false. The old variable period API will be removed.

**Comment by Kappie** [comment]
Thanks for your quick response. Any workaround that I can apply in the meantime?

**Comment by joaander** [comment]
You already identified the line of code that needs to be fixed - change `float` to `int`. If it is important that these values are not off by one, submit a PR against maint. I will continue to release minor bug fixes between now and the HOOMD 3.0 release.

Alternately, you can work around this by managing your `run()` steps in a loop. This example demonstrates the form of the loop, not the exact logrithmic spacing in your code.
```
obj = hoomd.some_periodic_object(period=1)
for i in range(10):
    obj.enable()
    run(1)
    obj.disable()
    run(i*1000-1)
```

**Comment by Kappie** [comment]
Thanks. I verified that changing lines 216 and 323 in System.h

```C++
int next = (int)pybind11::cast<float>(pynext) + m_created_tstep;
```
into 

```C++
int next = pybind11::cast<int>(pynext) + m_created_tstep;
```

fixes the problem. I will submit a PR later today.
